docs/exceptions.md

```diff
@@ -164,13 +195,29 @@ to return `Maybe v`. In 100% pure code, use one of these ways to handle errors:
 * `MaybeT`, `ExceptT e`
 * `CatchT`
 
+*Comment:* neither of these examples are of errors or exceptions! They are
+examples of normal values in the images of pure functions. The `Map` does not
+make any guarantee to that any particular key is in the map! Nor does the parser
+guarantee that every input will be recognized! These facts are reflected in the
+types. This has *nothing at all* to do with errors or exceptions. We need to
+strike this section out.
+
 Avoid using `Text` with the error message in place of `e` -- create a
 proper ADT. In case creating a proper ADT feels too cumbersome, use
 `CatchT`, which is equivalent to `ExceptT SomeException`. Note,
 however, that using `SomeException` in pure code is not the best
 practice, because the set of all possible exceptions is statically
 known. Use it only if you are lazy to define yet another ADT.
 
+*Comment:* nor is overusing exceptions the best practice, not the least because
+they do not appear in types, so we're essentially throwing away one of the
+strongest features of Haskell. Instead of conjuring up a new exception type and
+throwing it in the case of a completely normal condition, think about the
+contract that your function gives to its caller. Does the caller have a
+guarantee that this header hash picks out a block in the database? Of course
+not! So don't promise to give back a block for every input. That's not
```

Comment:
A type signature like `IO HeaderHash` communicates clearly that the header hash might not be in the database, because `IO` might throw, that's how `IO` works. In case you're worried that the possible exceptions are not listed in the type, well, I don't like it either, but we have to live with that or use a different base monad rather than IO (for example, https://hackage.haskell.org/package/unexceptionalio)

The issue is that in `ExceptT HeaderNotFound IO Header` we really have two ways to throw: `throwIO HeaderNotFound` and `throwError HeaderNotFound`. So at call site we must account for both exit paths when catching.

Reply:
> A type signature like IO HeaderHash communicates clearly that the header hash might not be in the database, because IO might throw, that's how IO works.

This type signature communicates that you can *always* get a `HeaderHash` using `IO`, unless there was an exception. Whereas, `HeaderHash -> IO (Either Oops Header)` communicates that even if the `IO` went smoothly with no exceptions, there still may not be a `Header` for that hash. The latter is what we want, when looking up a header hash.

> the possible exceptions are not listed in the type, well, I don't like it either, but we have to live with that or use a different base monad rather than IO 

No, there's a very simple solution: `HeaderHash -> IO (Either Oops Header)` where `Oops` includes all of the non-exceptional reasons why the `Header` could not be produced.

Reply:
>  includes all of the non-exceptional reasons 

* how do you decide which reasons are exceptional and which ones aren't?
* how do you deal with other libraries using `IO` exceptions to represent non-exceptional reasons something couldn't be done, including core libraries?

I'm ok with `Either Oops Header` as long as

* `Oops` doesn't have an `Exception` instance
* we're not using `ExceptT` internally to `throwError` this `Oops`, because `ExceptT` is for exceptional cases

Reply:
> how do you decide which reasons are exceptional and which ones aren't?

It's a good question. Imagine a perfect world where the definition could be given without any `IO`/side effects. What would the type be? 

> how do you deal with other libraries using IO exceptions to represent non-exceptional reasons something couldn't be done, including core libraries?

This would be dealt with internally. If you need to catch exceptions, catch them.

> we're not using ExceptT internally to throwError this Oops, because ExceptT is for exceptional cases

What does the caller care whether I use `ExceptT` internally? Is this property even observable by the caller? I don't think so. I guess this is just a code style thing?

Reply:
> This would be dealt with internally. If you need to catch exceptions, catch them.

But where do I catch them, every time I call some library function? There are really two options here:

* catch exceptions that represent normal scenarios every time we use a function from an external library; this basically leads us to having a ton of wrappers for every library function we might want to use

* catch exceptions when we want to handle them; this means that we end up with a mix of error handling techniques in our code, where we catch both `IO` exceptions and `ExceptT`-exceptions

Neither is satisfactory.

Reply:
> I guess this is just a code style thing?

We discussed this with @gromakovsky. We're going to amend the guidelines to allow `ExceptT E` over `IO` (or an abstract monad) as long as the `E` type has an instance that disallows throwing it as an `IO` exception:

```
instance TypeError "Not an exception" => Exception E
```

Reply:
@avieth https://github.com/input-output-hk/cardano-sl/pull/2507

Does this look better? You can keep your usage of `ExceptT`, just replace `Text` with an ADT.